components/data_prefetch/data_prefetch.tsx

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import PQueue from 'p-queue';
+import {Channel} from 'mattermost-redux/types/channels';
+
+import {Constants} from 'utils/constants';
+import {loadProfilesForSidebar} from 'actions/user_actions.jsx';
+
+const queue = new PQueue({concurrency: 2});
```

Comment:
Shouldn't `queue` be part of the `DataPrefetch` component? Having it be global feels weird.

Also, why do we have two different queues here, one in this file and one coming from redux?

Reply:
Actually, I don't understand what this queue even does. I see that stuff is added to it in this file, but I don't see what's being added actually get used for anything.

Reply:
We don't have a redux queue, It is just the queue state in redux which is used to not make multiple requests as messages comes into the channel or prevent duplicate requests in cases other components require it.

Reply:
Why not move the queue to be a field of the component though?

Reply:
Any particular reason why we want to move it to the component? I think it is a Usually way convention to initiate classes for external libraries, Why assign it to an instance variable in the component?

Things like this don't change in the lifetime of the app, so i don't see. a good reason to attach it to the instance unless it is a really big object that needs to be GC'ed by the browser.

Reply:
Avoiding global variables is generally considered to be a good practice since it prevents issues like multiple instances of DataPrefetch affecting each other's behaviour (if we ended up with multiple instances of it) and it helps prevents memory leaks by giving data an owner (like how Rust works without a garbage collector).

In this case, `queue` will be able to hold onto a references to an instance of `DataPrefetch` after it gets unmounted because the calls we're queuing directly reference ([like this one](https://github.com/mattermost/mattermost-webapp/pull/5547/files/e8634150fedef056f7ce3a8734baca9569ecd044#diff-799dd3cdcb584fe132d5d07f232327d0R79)) that hold a reference to the component.

Reply:
This is not a global variable. This is scoped inside a module. Like each file works as a module with our build systems and none of the values defined out of component are never global variables and are garbage collected when not needed. 

I don't think GC works that way in the case you described. They cause memory leaks only if they hang on to a detached node or held on with an eventlistner and that is not the case here. I remember seeing quite a few instances across open source projects where a new instance of object is created similar to this at the top of module as I said this is perfectly fine and indeed i say preferred if we don't need to change the way they are instantiated.

Reply:
`queue` can hold onto references of detached nodes, at least for a time while it's working through the queued. It's the same as those warnings we'd get by calling `setState` from async code after a component unmounted, although we just silenced those by checking if the component is mounted which doesn't actually solve that problem now that I think of it. Either way, those references don't hang around for that long, so this is fine for now, although we might want to keep an eye on it if memory usage starts spiking in the future.